services/src/main/java/com/linkedin/kmf/services/TopicRebalancer.java

```diff
@@ -0,0 +1,427 @@
+   *                                   rebalancePartitionFactor * brokerCount.
+   * @param topic Topic identifier
+   * @param zkConnect Zoo keeper connection url
+   * @param addPartitionCallback This gets called when partitions have been added to the monitored topic.
+   * @param scheduleDurationMs The duration between times this is scheduled to run in milliseconds.
+   * @param replicationFactor The desired replication factor when creating new partitions.
+   */
+  TopicRebalancer(double rebalanceThreshold, int rebalancePartitionFactor, String topic, String zkConnect,
+      PartitionsAddedCallback addPartitionCallback, int scheduleDurationMs,
+      int replicationFactor) {
+
+    if (rebalanceThreshold < 1) {
+      throw new IllegalArgumentException(
+          "rebalanceThreshold must be greater than or equal to one, but found " + rebalanceThreshold + ".");
+    }
+
+    if (addPartitionCallback == null) {
+      throw new NullPointerException("addPartitionCallback may not be null");
+    }
+
+    //TODO:  should there be more parameter checking or should I assume that ProduceService has done this.
+
+    _rebalanceThreshold = rebalanceThreshold;
+    _topic = topic;
+    _zkConnect = zkConnect;
+    _addPartitionCallback = addPartitionCallback;
+    _rebalancePartitionFactor = rebalancePartitionFactor;
+    _scheduleDurationMs = scheduleDurationMs;
+    _replicationFactor = replicationFactor;
+  }
+
+  int scheduleDurationMs() {
+    return _scheduleDurationMs;
+  }
+
+  @Override
+  public void run() {
+    try {
+      RebalanceCondition rebalanceCondition = monitoredTopicNeedsRebalance();
+      if (rebalanceCondition != RebalanceCondition.OK) {
+        LOG.info("Monitored topic \"" + _topic + "\" rebalance started.");
+        rebalanceMonitoredTopic(rebalanceCondition);
+        LOG.info("Monitored topic \"" + _topic + "\" rebalance complete.");
+      }
+
+    } catch (Exception e) {
+      LOG.error("Monitored topic rebalance failed with exception.", e);
+    }
+  }
+
+  /**
+   * <pre>
+   * Each time this is invoked one of the following happens.  This is done because I don't have a way to wait for any
+   * of these steps to complete.
+   *
+   * 1. if number of partitions falls below threshold then create new partitions
+   *     create new produce runnables (callback does this)
+   *     create new metrics (callback does this)
+   *    increment _PartitionNum (callback does this)
+   * 2. This is done during rebalance.
+   *   if a broker does not have enough partitions then assign it partitions from brokers that have excess
+   *   if a broker is not a leader of some partition then make it a leader of some partition by finding out
+   *     which broker is a leader of more than one partition.  Does this involve moving partitions among brokers?
+   * 3. run a preferred replica election
+   * </pre>
+   */
+  void rebalanceMonitoredTopic(RebalanceCondition rebalanceCondition) {
+    ZkUtils zkUtils = createZooKeeperUtils();
```

Comment:
It seems that we zkUtils will be created exactly once. Can we simply create it once in the constructor? If we are concerned that this is a blocking call, then we can create it in the method `Service.start()`.